src/charm.py

```diff
@@ -128,317 +248,235 @@ def set_pod_spec(self, event):
+        }
+
+        return Layer(layer_config)
+
+    #
+    # Check if connection can be made with container
+    #
+    def _check_container_connection(self):
+        if not self.container.can_connect():
+            raise CheckFailed("Pod startup is not complete", MaintenanceStatus)
 
+    #
+    # Check for leader
+    #
     def _check_leader(self):
         if not self.unit.is_leader():
-            log.info("Not a leader, skipping set_pod_spec")
+            self.logger.info("Not a leader, skipping setup")
             raise CheckFailed("Waiting for leadership", WaitingStatus)
 
+    #
+    # Update Pebble configuration layer if changed.
+    #
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer if changed."""
+        current_layer = self.container.get_plan()
+        new_layer = self._seldon_core_operator_layer
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self.container.add_layer(self._container_name, new_layer, combine=True)
+            try:
+                self.logger.info(
+                    "Pebble plan updated with new configuration, replaning"
+                )
+                self.container.replan()
+            except ChangeError:
+                raise CheckFailed("Failed to replan", BlockedStatus)
 
-class CheckFailed(Exception):
-    """Raise this exception if one of the checks in main fails."""
+    #
+    # Upload certs to container
+    #
+    def _upload_certs_to_container(self):
 
-    def __init__(self, msg, status_type=None):
-        super().__init__()
+        self.container.push(
+            "/tmp/k8s-webhook-server/serving-certs/tls.key",
+            self._stored.key,
+            make_dirs=True,
+        )
+        self.container.push(
+            "/tmp/k8s-webhook-server/serving-certs/tls.crt",
+            self._stored.cert,
+            make_dirs=True,
+        )
+        self.container.push(
+            "/tmp/k8s-webhook-server/serving-certs/ca.crt",
+            self._stored.ca,
+            make_dirs=True,
+        )
 
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
+    #
+    # Deploy all K8S resouces
+    #
+    def _deploy_k8s_resources(self) -> None:
+        try:
+            self.unit.status = MaintenanceStatus("Creating K8S resources")
+            self.k8s_resource_handler.apply()
+            self.crd_resource_handler.apply()
+            self.configmap_resource_handler.apply()
+        except ApiError:
+            raise CheckFailed("K8S resources creation failed", BlockedStatus)
+        self.model.unit.status = MaintenanceStatus("K8S resources created")
+
+    #
+    # Main entry point for the Charm
+    #
+    def main(self, _) -> None:
+
+        try:
+            self._check_container_connection()
+            self._check_leader()
+            self._upload_certs_to_container()
```

Comment:
is there any downside to uploading these on all hooks rather than just at install time?  I don't think there is, but just asking

Reply:
Yes, you are correct. Those are generated at init time only. Might as well upload those then.

Reply:
Done.